File: 1340. jump-game-v.py
class Solution:
    def maxJumps(self, arr: list[int], d: int) -> int:
        n = len(arr)
        dp = [-1] * n

        def dfs(i):
            if dp[i] != -1:
                return dp[i]

            best = 1

            # Right scan
            for nxt in range(i + 1, min(n, i + d + 1)):
                if arr[nxt] >= arr[i]:
                    break

                best = max(best, 1 + dfs(nxt))

            # Left scan
            for nxt in range(i - 1, max(-1, i - d - 1), -1):
                if arr[nxt] >= arr[i]:
                    break

                best = max(best, 1 + dfs(nxt))

            dp[i] = best
            return dp[i]

        return max(dfs(i) for i in range(n))

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    arr1 = [6,4,14,6,8,13,9,7,10,6,12]
    d1 = 2
    output1 = sol.maxJumps(arr1, d1)
    print(f"Example 1: arr={arr1}, d={d1}")
    print(f"Output: {output1}")
    print(f"Expected: 4")
    print()
    
    # Example 2
    arr2 = [3,3,3,3,3]
    d2 = 3
    output2 = sol.maxJumps(arr2, d2)
    print(f"Example 2: arr={arr2}, d={d2}")
    print(f"Output: {output2}")
    print(f"Expected: 1")
    print()
    
    # Example 3
    arr3 = [7,6,5,4,3,2,1]
    d3 = 1
    output3 = sol.maxJumps(arr3, d3)
    print(f"Example 3: arr={arr3}, d={d3}")
    print(f"Output: {output3}")
    print(f"Expected: 7")

File: 1340. jump-game-v.java
class Solution {
    int[] dp;

    public int maxJumps(int[] arr, int d) {
        int n = arr.length;
        dp = new int[n];

        int ans = 1;

        for (int i = 0; i < n; i++) {
            ans = Math.max(ans, dfs(i, arr, d));
        }

        return ans;
    }

    private int dfs(int i, int[] arr, int d) {
        if (dp[i] != 0)
            return dp[i];

        int best = 1;

        for (int nxt = i + 1; nxt <= Math.min(arr.length - 1, i + d); nxt++) {
            if (arr[nxt] >= arr[i])
                break;

            best = Math.max(best, 1 + dfs(nxt, arr, d));
        }

        for (int nxt = i - 1; nxt >= Math.max(0, i - d); nxt--) {
            if (arr[nxt] >= arr[i])
                break;

            best = Math.max(best, 1 + dfs(nxt, arr, d));
        }

        return dp[i] = best;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        int[] arr1 = {6,4,14,6,8,13,9,7,10,6,12};
        int d1 = 2;
        int output1 = sol.maxJumps(arr1, d1);
        System.out.println("Example 1: arr=[6,4,14,6,8,13,9,7,10,6,12], d=2");
        System.out.println("Output: " + output1);
        System.out.println("Expected: 4");
        System.out.println();
        
        // Example 2
        int[] arr2 = {3,3,3,3,3};
        int d2 = 3;
        int output2 = sol.maxJumps(arr2, d2);
        System.out.println("Example 2: arr=[3,3,3,3,3], d=3");
        System.out.println("Output: " + output2);
        System.out.println("Expected: 1");
        System.out.println();
        
        // Example 3
        int[] arr3 = {7,6,5,4,3,2,1};
        int d3 = 1;
        int output3 = sol.maxJumps(arr3, d3);
        System.out.println("Example 3: arr=[7,6,5,4,3,2,1], d=1");
        System.out.println("Output: " + output3);
        System.out.println("Expected: 7");
    }
}